src/entities/nestedExit/getQueryCallsAttributes.ts

```diff
@@ -0,0 +1,201 @@
+import { TokenAmount } from '../tokenAmount';
+import { Address } from '../../types';
+
+export const getQueryCallsAttributes = (
+    input: NestedProportionalExitInput | NestedSingleTokenExitInput,
+    { pools }: NestedPoolState,
+    isProportional: boolean,
+): {
+    bptAmountIn: TokenAmount;
+    callsAttributes: NestedExitCallAttributes[];
+} => {
+    const {
+        bptAmountIn,
+        chainId,
+        accountAddress,
+        useNativeAssetAsWrappedAmountOut = false,
+        toInternalBalance = false,
+    } = input;
+    let callsAttributes: NestedExitCallAttributes[];
+
+    // sort pools by descending level
+    const poolsTopDown = pools.sort((a, b) => b.level - a.level);
+
+    if (isProportional) {
+        /**
+         * Overall logic to build sequence of proportional exit calls:
+         * 1. Go from top pool to bottom filling out input amounts and output refs
+         * 2. Inputs will be bptAmountIn provided or output of the previous level
+         * 3. Output at bottom level is the amountsOut
+         */
+
+        callsAttributes = getProportionalExitCallsAttributes(
+            poolsTopDown,
+            chainId,
+            useNativeAssetAsWrappedAmountOut,
+            accountAddress,
+            bptAmountIn,
+            toInternalBalance,
+        );
+    } else {
+        /**
+         * Overall logic to build sequence of single token exit calls:
+         * 1. Go BOTTOM-UP building exit path to tokenOut
+         * 2. Go through exit path filling out input amounts and output refs
+         * 3. Inputs will be bptAmountIn provided or output of the previous level
+         * 4. Output at bottom level is the amountOut
+         */
+
+        const { tokenOut } = input as NestedSingleTokenExitInput;
+
+        callsAttributes = getSingleTokenExitCallsAttributes(
+            poolsTopDown,
+            chainId,
+            useNativeAssetAsWrappedAmountOut,
+            accountAddress,
+            bptAmountIn,
+            toInternalBalance,
+            tokenOut,
+        );
+    }
+
+    const bptIn = new Token(chainId, poolsTopDown[0].address, 18);
+    const _bptAmountIn = TokenAmount.fromRawAmount(bptIn, bptAmountIn);
+    return { callsAttributes, bptAmountIn: _bptAmountIn };
+};
+
+export const getProportionalExitCallsAttributes = (
+    poolsSortedByLevel: NestedPool[],
+    chainId: number,
+    useNativeAssetAsWrappedAmountOut: boolean,
+    accountAddress: Address,
+    bptAmountIn: bigint,
+    toInternalBalance: boolean,
+) => {
+    const calls: NestedExitCallAttributes[] = [];
+    for (const pool of poolsSortedByLevel) {
+        const sortedTokens = pool.tokens
+            .sort((a, b) => a.index - b.index)
+            .map((t) => new Token(chainId, t.address, t.decimals));
+
+        const sortedTokensWithoutBpt = sortedTokens.filter(
+            (t) => !t.isSameAddress(pool.address),
+        );
+        const upperLevelCall = calls.find((call) =>
```

Comment:
Is `upperLevel` right here?

Reply:
It's actually trying to find a previous call that has its output as the current call input.
We could have a more descriptive name, such as `previousCall` or `previousNestedCall`
Or if that's too complex to describe on a variable name, we could add more details on a comment.

Reply:
I renamed this upperLevelCall to `previousCall` and refactored it into `getBptAmountIn` logic to make it easier to understand. Let me know what you think.

Reply:
Think that helps a lot!